pkg/netflow/portrollup/portrolluptracker.go

```diff
@@ -0,0 +1,54 @@
+// Unless explicitly stated otherwise all files in this repository are licensed
+// under the Apache License Version 2.0.
+// This product includes software developed at Datadog (https://www.datadoghq.com/).
+// Copyright 2022-present Datadog, Inc.
+
+package portrollup
+
+type portRollupTracker struct {
+	// represent: map[<SOURCE_PORT>]map[<DEST_PORT>]true
+	sourcePorts map[uint16]map[uint16]bool
+
+	// represent: map[<DEST_PORT>]map[<SOURCE_PORT>]true
+	destPorts map[uint16]map[uint16]bool
+}
+
+func newPortRollupTracker() *portRollupTracker {
+	return &portRollupTracker{
+		sourcePorts: make(map[uint16]map[uint16]bool),
+		destPorts:   make(map[uint16]map[uint16]bool),
+	}
+}
+
+func (pr *portRollupTracker) getSourcePortCount(port uint16) uint16 {
+	return uint16(len(pr.sourcePorts[port]))
+}
+
+func (pr *portRollupTracker) getDestPortCount(port uint16) uint16 {
+	return uint16(len(pr.destPorts[port]))
+}
+
+func (pr *portRollupTracker) add(sourcePort uint16, destPort uint16, portRollupThreshold int) {
+	// we only track additional ports if portRollupThreshold has not been reached
+	// if portRollupThreshold is already reached, there is no value to track more ports
+	// since we already know we need to rollup
+
+	sourceToDestPorts := len(pr.sourcePorts[sourcePort])
+	destToSourcePorts := len(pr.destPorts[destPort])
+	if sourceToDestPorts >= portRollupThreshold || destToSourcePorts >= portRollupThreshold {
+		return
+	}
+
+	addPort(pr.sourcePorts, sourcePort, destPort, portRollupThreshold)
+	addPort(pr.destPorts, destPort, sourcePort, portRollupThreshold)
+}
+
+func addPort(ports map[uint16]map[uint16]bool, port uint16, ephemeralPort uint16, portRollupThreshold int) {
+	if _, ok := ports[port]; !ok {
+		ports[port] = make(map[uint16]bool)
+	}
+
+	if len(ports[port]) < portRollupThreshold && !ports[port][ephemeralPort] {
```

Comment:
<!-- Questions are appropriate if you have a potential concern but are not quite sure if it's relevant or not. Asking the author for clarification or investigation can lead to a quick resolution. -->
**❓ question**
Shouldn't this be `len(ports[port]) > portRollupThreshold` ?

Reply:
It will work too, but I would like to avoid writing `ports[port][ephemeralPort] = true` when the value is already `true`, hence the use of `&& !ports[port][ephemeralPort]`.

Reply:
Let me rephrase, shouldn't this be 
```suggestion
	if len(ports[port]) > portRollupThreshold && !ports[port][ephemeralPort] {
```

Reply:
I see, thanks for rephrasing. It's actually related to:

> 	// we only track additional ports if portRollupThreshold has not been reached
>	// if portRollupThreshold is already reached, there is no value to track more ports
>	// since we already know we need to rollup

But since the check is already made here: https://github.com/DataDog/datadog-agent/pull/13367/files/8000e6a18acce8b6ecd55f6da5509ac09828d348#diff-544bcf72f2eff2bc13439925d88e223b59ba3df2e265388bbeddee5041eb4d9aR38-R40

There is no need to do an additional check here: Removed here https://github.com/DataDog/datadog-agent/pull/13367/commits/a044fea914964b59f13a07b1246b5015ccdfd253